In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn import linear_model


In [168]:
df = pd.read_csv('wc-wo-outliers.csv')

In [169]:
new_X = df[['goals_z', 'xg_z', 'nsxg_z','clearances_z', 
              'progpasses_z', 'interceptions_z']]

In [170]:
new_X

goals_z      xg_z    nsxg_z  clearances_z  progpasses_z  interceptions_z
0    0.423077  0.146923 -0.124615     -0.076923      0.037692         0.270000
1    0.216923  0.348462  0.146923     -0.059231      0.626923         0.834615
2    0.113846  0.392308  0.786923      0.084615      0.405385         0.859231
3    0.479231  0.609231  0.692308      0.020769      0.042308         1.015385
4    0.877692  0.773846  0.890769     -0.117692      0.335385         0.637692
..        ...       ...       ...           ...           ...              ...
143 -0.261538  0.207692  0.005385     -0.279231      0.176154        -0.169231
144  0.234615  0.405385  0.296154     -0.392308      0.331538         0.497692
145  0.536154  0.170769  0.228462      0.177692     -0.196154         0.343077
146 -0.070769 -0.368462 -0.526154     -0.313846     -0.536923        -0.622308
147  0.472308  0.458462  0.521538     -0.239231      0.273846         0.178462

[148 rows x 6 columns]

In [171]:
y = df['results']

X_tr, X_tt, y_tr, y_tt = train_test_split(
     new_X, y, test_size=0.20, random_state=1)

In [165]:
y = df['results']
X = df.drop(columns=['results'])

X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=1)

In [166]:
reg = linear_model.LinearRegression()
reg.fit(X_train, y_train)
pred_train= reg.predict(X_train)

In [172]:
reg = linear_model.LinearRegression()
reg.fit(X_tr, y_tr)
pred_new= reg.predict(X_tr)

Theres no clear difference between using all the features or a few features. mape goes back and forth depending on the random_state value

In [167]:
print('Linear Regression ')
print('---------------------')
print('RMSE: '+str(np.sqrt(mean_squared_error(y_train,pred_train))))
print('R Squared: '+str(r2_score(y_train, pred_train)))
print('MSE: '+str(mean_squared_error(y_train, pred_train)))
print('MAE: '+str(mean_absolute_error(y_train, pred_train)))
print('MAPE: '+str(mean_absolute_percentage_error(y_train, pred_train)))

print('---------------------')

pred_test= reg.predict(X_test)

print('RMSE : '+str(np.sqrt(mean_squared_error(y_test,pred_test)))) 
print('R2 : '+str(r2_score(y_test, pred_test)))
print('MSE: '+str(mean_squared_error(y_test, pred_test)))
print('MAE: '+str(mean_absolute_error(y_test, pred_test)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, pred_test)))

Linear Regression 
---------------------
RMSE: 6.322750364864893
R Squared: 0.5388878434290846
MSE: 39.97717217639914
MAE: 5.0711348407757555
MAPE: 0.9277204828534783
---------------------
RMSE : 7.589107793384392
R2 : 0.2950386733173721
MSE: 57.59455709960772
MAE: 6.294988780232196
MAPE: 0.8948800905643852


In [173]:
print('Linear Regression ')
print('---------------------')
print('RMSE: '+str(np.sqrt(mean_squared_error(y_tr,pred_new))))
print('R Squared: '+str(r2_score(y_tr, pred_new)))
print('MSE: '+str(mean_squared_error(y_tr, pred_new)))
print('MAE: '+str(mean_absolute_error(y_tr, pred_new)))
print('MAPE: '+str(mean_absolute_percentage_error(y_tr, pred_new)))

print('---------------------')

pred_tt= reg.predict(X_tt)

print('RMSE : '+str(np.sqrt(mean_squared_error(y_tt,pred_tt)))) 
print('R2 : '+str(r2_score(y_tt, pred_tt)))
print('MSE: '+str(mean_squared_error(y_tt, pred_tt)))
print('MAE: '+str(mean_absolute_error(y_tt, pred_tt)))
print('MAPE: '+str(mean_absolute_percentage_error(y_tt, pred_tt)))

Linear Regression 
---------------------
RMSE: 6.621010217303568
R Squared: 0.49435814322527516
MSE: 43.83777629763824
MAE: 5.358710334963796
MAPE: 0.9769333374230189
---------------------
RMSE : 7.14723851068569
R2 : 0.37474035420356944
MSE: 51.083018328628604
MAE: 6.151067059736851
MAPE: 0.8645396645603936


In [6]:
x_train = X_train[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [7]:
x_test = X_test[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [8]:
reg.fit(x_train, y_train)
pred_train= reg.predict(x_train)

In [9]:
print('Linear Regression ')
print('---------------------')
print('RMSE: '+str(np.sqrt(mean_squared_error(y_train,pred_train))))
print('R Squared: '+str(r2_score(y_train, pred_train)))
print('MSE: '+str(mean_squared_error(y_train, pred_train)))
print('MAE: '+str(mean_absolute_error(y_train, pred_train)))
print('MAPE: '+str(mean_absolute_percentage_error(y_train, pred_train)))

print('---------------------')

pred_test= reg.predict(x_test)

print('RMSE : '+str(np.sqrt(mean_squared_error(y_test,pred_test)))) 
print('R2 : '+str(r2_score(y_test, pred_test)))
print('MSE: '+str(mean_squared_error(y_test, pred_test)))
print('MAE: '+str(mean_absolute_error(y_test, pred_test)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, pred_test)))

Linear Regression 
---------------------
RMSE: 6.653300735302932
R Squared: 0.48941410864051593
MSE: 44.266410674382534
MAE: 5.4312779157090025
MAPE: 0.9624719869928929
---------------------
RMSE : 6.901090797137889
R2 : 0.41706607228022774
MSE: 47.62505419034126
MAE: 5.974021049665942
MAPE: 0.8674060149631102


the lowest MAPE value, which tells us that it’s able to forecast the most accurately among the potential models.

In [10]:
from sklearn.metrics import mean_squared_log_error
print('Mean Squared Log Error: ' +str(mean_squared_log_error( y_test, pred_test, squared=False)))

Mean Squared Log Error: 0.5892030828713158


In [12]:
# #display adjusted R-squared
# print('Adjusted r^2: ' +str(1 - (1-reg.score(X_train, y_train))*(len(y_train)-1)/(len(y_train)-X_train.shape[1]-1)))

The value for R-squared can range from 0 to 1. A value of 0 indicates that the response variable cannot be explained by the predictor variable at all while a value of 1 indicates that the response variable can be perfectly explained without error by the predictor variables.

- AIC stands for (Akaike’s Information Criteria), a metric developped by the Japanese Statistician, Hirotugu Akaike, 1970. The basic idea of AIC is to penalize the inclusion of additional variables to a model. It adds a penalty that increases the error when including additional terms. The lower the AIC, the better the model.
- AICc is a version of AIC corrected for small sample sizes.
- BIC (or Bayesian information criteria) is a variant of AIC with a stronger penalty for including additional variables to the model.
- Mallows Cp: A variant of AIC developed by Colin Mallows.


In [13]:
import statsmodels.api as sm

In [14]:
#add constant to predictor variables
X_train = sm.add_constant(X_train)

#fit regression model
model = sm.OLS(y_train, X_train).fit()

#view AIC of model
print('AIC: '+str(model.aic))

AIC: 804.0899081188967


In [15]:
X_train.head()

const   goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
126    1.0 -0.018462  0.184615   0.168462      0.276923 -0.106923   
97     1.0  0.198462 -0.116154  -0.090000     -0.013077  0.193077   
42     1.0  0.214615  0.523077  -0.176154      0.277692 -0.196923   
135    1.0  0.766923  0.643846  -0.265385      0.230000  0.353077   
28     1.0  0.332308  0.362308  -0.043846      0.316923  0.060000   

     progpasses_z  takeons_z  progruns_z  tackles_z  interceptions_z  \
126     -0.225385   0.138462    0.063846  -0.024615        -0.153846   
97       0.132308   0.010769    0.395385   0.102308         0.802308   
42      -0.245385  -0.013846   -0.178462  -0.228462        -0.250000   
135     -0.015385   0.496154    0.606923   0.446154        -0.096923   
28       0.019231   0.140769    0.210000   0.343846         0.276154   

     clearances_z  blocks_z  aerials_z   fouls_z  fouled_z    nsxg_z  
126     -0.119231  0.056923   0.285385 -0.036923 -0.293077  0.228462  
97      -0.154615 -0.013077   0.016154  0.668462  0.585385 -0.030000  
42       0.096154  0.032308   0.760769 -0.122308  0.119231  0.124615  
135     -0.128462  0.086923  -0.233846  0.180769 -0.118462  0.404615  
28      -0.057692  0.112308  -0.053846  0.143077  0.060769  0.192308

In [16]:
new_x = X_train[['goals_z', 'boxtouches_z', 'fouls_z', 'fouled_z', 'blocks_z']]

In [17]:
#add constant to predictor variables
new_x = sm.add_constant(new_x)

#fit regression model
model = sm.OLS(y_train, new_x).fit()

#view AIC of model
print(model.aic)

794.4357164871906


for AIC - the model with a lower AIC value is the better fitting model. 

https://www.statology.org/aic-in-python/

In [18]:
x = X_train[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [19]:
#add constant to predictor variables
x = sm.add_constant(x)

#fit regression model
model = sm.OLS(y_train, x).fit()

#view AIC of model
print(model.aic)

792.1161813534175


In [20]:
x_t = X_test[['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']]

In [21]:
#add constant to predictor variables
x_t = sm.add_constant(x_t)

#fit regression model
model = sm.OLS(y_test, x_t).fit()

#view AIC of model
print(model.aic)

209.67105935243535


In [22]:
#add constant to predictor variables
x = sm.add_constant(X_test)

#fit regression model
model = sm.OLS(y_test, x).fit()

#view AIC of model
print(model.aic)

222.00398789642531


In [27]:
df.columns

Index(['goals_z', 'xg_z', 'crosses_z', 'boxtouches_z', 'passes_z',
       'progpasses_z', 'takeons_z', 'progruns_z', 'tackles_z',
       'interceptions_z', 'clearances_z', 'blocks_z', 'aerials_z', 'fouls_z',
       'fouled_z', 'nsxg_z', 'results'],
      dtype='object')

In [86]:
x_t = X_test[['goals_z', 'xg_z', 'nsxg_z','clearances_z', 
              'progpasses_z', 'interceptions_z']]

In [87]:
#add constant to predictor variables
x_t = sm.add_constant(x_t)

#fit regression model
model = sm.OLS(y_test, x_t).fit()

#view AIC of model
print(model.aic)

205.04062985068555


- <b>205.0406 = ['goals_z', 'xg_z', 'nsxg_z','clearances_z', 'progpasses_z', 'interceptions_z']</b>
- 209.6710 = ['goals_z', 'xg_z', 'tackles_z', 'nsxg_z']
- 222.0039 = all features

# Feature Selection

In [ ]:
from sklearn.feature_selection import SelectKBest, chi2


# https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics

In [ ]:
from sklearn.metrics import explained_variance_score

In [ ]:
explained_variance_score(y_test, pred_test)

In [ ]:
from sklearn.metrics import mean_pinball_loss
mean_pinball_loss(y_test, pred_test, alpha=0.1)

The pinball loss output is a non-negative floating point. The best value is 0.0.

In [ ]:
from sklearn.metrics import mean_pinball_loss
mean_pinball_loss(y_test, pred_test, alpha=0.9)

https://www.kaggle.com/code/satishgunjal/tutorial-k-fold-cross-validation

In [ ]:
import numpy as np
from scipy.stats import ttest_ind

v1 = np.random.normal(size=100)
v2 = np.random.normal(size=100)

res = ttest_ind(v1, v2)

print(res)

In [ ]:
import numpy as np
from scipy.stats import ttest_ind

v1 = y_test
v2 = pred_test

res = ttest_ind(v1, v2)

print(res)

In [ ]:
from scipy.stats import kstest

v = y_train

res = kstest(v, 'norm')

print(res)

In [ ]:
from sklearn.model_selection import KFold, StratifiedKFold, cross_val_score
from sklearn import linear_model, tree, ensemble

In [ ]:
# Lets split the data into 5 folds.  
# We will use this 'kf'(KFold splitting stratergy) object as input to cross_val_score() method
kf =KFold(n_splits=5, shuffle=True, random_state=42)

cnt = 1
# split()  method generate indices to split data into training and test set.
for train_index, test_index in kf.split(X, y):
    print(f'Fold:{cnt}, Train set: {len(train_index)}, Test set:{len(test_index)}')
    cnt += 1



In [ ]:
def rmse(score):
    rmse = np.sqrt(-score)
    print(f'rmse= {"{:.2f}".format(rmse)}')

In [ ]:
score = cross_val_score(linear_model.LinearRegression(), X, y, cv= kf, scoring="neg_mean_squared_error")
print(f'Scores for each fold: {score}')
rmse(score.mean())

In [ ]:
score = cross_val_score(tree.DecisionTreeRegressor(random_state= 42), X, y, cv=kf, scoring="neg_mean_squared_error")
print(f'Scores for each fold: {score}')
rmse(score.mean())

In [ ]:
score = cross_val_score(ensemble.RandomForestRegressor(random_state= 42), X, y, cv= kf, scoring="neg_mean_squared_error")
print(f'Scores for each fold are: {score}')
rmse(score.mean())

In [ ]:
max_depth = [1,2,3,4,5,6,7,8,9,10]

for val in max_depth:
    score = cross_val_score(tree.DecisionTreeRegressor(max_depth= val, random_state= 42), X, y, cv= kf, scoring="neg_mean_squared_error")
    print(f'For max depth: {val}')
    rmse(score.mean())

In [ ]:
estimators = [50, 100, 150, 200, 250, 300, 350]

for count in estimators:
    score = cross_val_score(ensemble.RandomForestRegressor(n_estimators= count, random_state= 42), X, y, cv= kf, scoring="neg_mean_squared_error")
    print(f'For estimators: {count}')
    rmse(score.mean())